src/python/torchdistx/gossip_grad.py

```diff
@@ -0,0 +1,338 @@
+        torch.cuda.set_device(self.rank)
+
+        # Master worker for a current local `process_group`
+        self.master_worker = c10d._get_global_rank(self.local_process_group, 0)
+
+    def _create_master_group(self):
+        r"""
+        Creates master process group, i.e. a group of workers,
+        which communicate gradients between different nodes.
+        """
+        # Every 0th worker on every node will be assigned to a master group,
+        # i.e. if number of rocesses per node is 8, master group contains
+        # 0th, 8th, 16th, 24th, 32nd, ... ranks
+        ranks = [i * self.proc_per_node for i in range(self.num_nodes)]
+        return dist.new_group(ranks)
+
+    def _generate_topologies(self):
+        r"""
+        Creates `num_nodes` random topology shuffles and returns an infinite iterator.
+        Original topology is of the form:
+            [0*K, 1*K, ... , N*K],
+        where N is the number of nodes and K - the number of workers on each node.
+        For example, with N=4 and K=8, original topology is
+            [0, 8, 16, 24]
+
+        Workers' rank values are used instead of node values for easier peer assignment
+        in a collective communication stage.
+
+        Returns:
+            An infinite iterator over created topologies
+        """
+        random.seed(self.num_nodes * 100)
+        topologies_set = []
+        original_list = [i * self.proc_per_node for i in range(self.num_nodes)]
+        for _ in range(self.num_nodes):
+            random.shuffle(original_list)
+            topologies_set.append(original_list.copy())
+
+        return cycle(topologies_set)
+
+
+def _get_send_recv_peers(state):
+    r"""
+    Computes peers for the collective communication stage.
+    For a CUBE topology a node sends grads to and receives from
+    the same neighboring vertex. A pick for a neighboring vertex
+    depends on the step number and current virtual topology in use.
+
+    For a DISSEMINATION topology a node typically sends grads
+    to and receives from different neighbors, but there may be a step
+    where send and receive peers are the same node. A pick for send and receive peers
+    depends on the step number and current virtual topology in use.
+
+    For more information, please refer to the original
+    `paper <https://arxiv.org/abs/1803.05880>`_
+
+    Args:
+        state (GossipGradState): State for GossipGraD communication hook.
+
+    Returns:
+        Peers' global ranks to whom a current node sends gradients
+        and from whom it is received.
+    """
+    assert state.gossip_period > 0, "`gossip_period` should be greater than 0."
+    power = state.iter % state.gossip_period
+    # Our new node_rank is a position of a global rank in
+    # a virtual topology
+    node_rank = state.cur_topology.index(state.rank)
+
+    if state.topology == Topology.CUBE:
+        peer_idx = node_rank ^ 2**power
+        if peer_idx >= len(state.cur_topology):
+            return -1, -1
```

Comment:
@rohan-varma this tracks situation for `CUBE` topology, when we have the number of communication peers not equal to a power of 2. Suppose instead of 8 nodes, we only have 6 (0, ..., 5). Then, at some point node 5 has to communicate with node 7. Node 7 is non-existing, thus I return -1 and during the gossiping stage (L243 -L244) we exit early and skip communication, since peer is not present.